src/main/java/appeng/parts/misc/ItemHandlerAdapter.java

```diff
@@ -0,0 +1,362 @@
+		ItemStack gathered = null;
+
+		final boolean simulate = ( mode == Actionable.SIMULATE );
+
+		for( int i = 0; i < itemHandler.getSlots(); i++ )
+		{
+			ItemStack sub = itemHandler.getStackInSlot( i );
+
+			if( !Platform.isSameItem( sub, req ) )
+			{
+				continue;
+			}
+
+			ItemStack extracted;
+
+			// We have to loop here because according to the docs, the handler shouldn't return a stack with size > maxSize, even if we
+			// request more. So even if it returns a valid stack, it might have more stuff.
+			do
+			{
+				extracted = itemHandler.extractItem( i, remainingSize, simulate );
+				if( extracted != null )
+				{
+					if( extracted.stackSize > remainingSize )
+					{
+						// Something broke. It should never return more than we requested... We're going to silently eat the remainder
+						AELog.warn( "Mod that provided item handler {} is broken. Returned {} items, even though we requested {}.",
+								itemHandler.getClass().getSimpleName(), extracted.stackSize, remainingSize );
+						extracted.stackSize = remainingSize;
+					}
+
+					// We're just gonna use the first stack we get our hands on as the template for the rest
+					if( gathered == null )
+					{
+						gathered = extracted;
+					}
+					else
+					{
+						gathered.stackSize += extracted.stackSize;
+					}
+					remainingSize -= gathered.stackSize;
+				}
+			}
+			while( extracted != null && remainingSize > 0 );
+
+			// Done?
+			if( remainingSize <= 0 )
+			{
+				break;
+			}
+		}
+
+		if( gathered != null )
+		{
+			if( mode == Actionable.MODULATE )
+			{
+				this.onTick();
+			}
+
+			return AEItemStack.create( gathered );
+		}
+
+		return null;
+	}
+
+	@Override
+	public TickRateModulation onTick()
+	{
+		LinkedList<IAEItemStack> changes = new LinkedList<IAEItemStack>();
+
+		int slots = itemHandler.getSlots();
+
+		// Make room for new slots
+		if( slots > cachedStacks.length )
+		{
+			cachedStacks = Arrays.copyOf( cachedStacks, slots );
+			cachedAeStacks = Arrays.copyOf( cachedAeStacks, slots );
+		}
+
+		for( int slot = 0; slot < slots; slot++ )
+		{
+			// Save the old stuff
+			ItemStack oldIS = cachedStacks[slot];
+			IAEItemStack oldAeIS = cachedAeStacks[slot];
+
+			ItemStack newIS = itemHandler.getStackInSlot( slot );
+
+			if( this.isDifferent( newIS, oldIS ) )
+			{
+				// Completely different item
+				cachedStacks[slot] = newIS;
+				cachedAeStacks[slot] = AEItemStack.create( newIS );
+
+				// If we had a stack previously in this slot, notify the newtork about its disappearance
+				if( oldAeIS != null )
+				{
+					oldAeIS.setStackSize( -oldAeIS.getStackSize() );
+					changes.add( oldAeIS );
+				}
+
+				// Notify the network about the new stack. Note that this is null if newIS was null
+				if( cachedAeStacks[slot] != null )
+				{
+					changes.add( cachedAeStacks[slot] );
+				}
+			}
+			else
+			{
+				// Still the same item, but amount might have changed
+				int newSize = ( newIS == null ? 0 : newIS.stackSize );
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove those useless parentheses. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'newSize' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)